### DISCUSSION TITLE
AppCast <sparkle:minimumSystemVersion> for different operating systems or enclosures

### DISCUSSION THREAD

**Discussion by MichaelKloskeZynaptiq** [sparkle-project/Sparkle#2580] [General]
Hi,
we are using Sparkle for our macOS versions of our applications and now we start using WinSparkle for the windows versions as well. Sparkle supports this by having a \<sparkle:os="os_name"\> - option in the \<enclusore\> tag. WinSparkle detects its enclosure and downloads it as expected.
But our mac apps require a minimum operating system of 10.15, therefore we added \<sparkle:minimumSystemVersion\>10.15\</sparkle:minimumSystemVersion\> to the appcast \<item\>.

Now, WinSparkle checks for minimum Windows version 10.15 where this flag is meant for macOS and would at least be different for Windows. I first tried to add the \<os\> - attribute in the \<minimumSystemVersion\> - tag but whether WinSparkle does not implement that for now or it is not the way it is meant to work.

Now I wanted to ask if the \<minimumSystemVersion\> would not be more useful if it was an attribute of the \<enclosure\> tag.
Otherwise we have to create always two items, one for windows and one for macOS with the exact same content except the enclosure.

**Comment by zorgiepoo** [comment]
Unfortunately, `sparkle:os` was not carefully designed, at least for the Mac version. I recommend using separate feeds, or like you said, two different items. Sparkle Mac does not actually support a single appcast item being shared by different OS's (it either thinks an appcast item is macOS or not macOS). I believe this may otherwise work by "accident" depending on the order of the enclosure items.

Also an appcast item does not strictly need enclosure items for info-only updates (which you cannot mark the os), and an appcast item can have multiple enclosure items for delta updates (therefore, minimumSystemVersion not being a part of the enclosure item is _correct._). To properly support "sparkle:os" it should be a supported attribute for every kind of element. But that kind of customization may be complicated and at that point you can use different items.

cc @vslavik

**Reply by MichaelKloskeZynaptiq** [reply]
From your points it is indeed much easier to just have two different feeds or items.

Thanks for your quick reply.

**Reply by zorgiepoo** [reply]
I recommend different feeds. I do not think having two different appcast items for each Mac/Windows version in the same appcast file is a notable simplification. Meanwhile, it may discourage people from using generate_appcast or generate deltas. And it may get awkward if you need to make an informational update without a download, like in case something gets hosed.

**Reply by vslavik** [reply]
I agree with the assessment. I'm somewhat confused about whether Sparke's code reads `os` from enclosure or item, but in any case, it definitely doesn't filter every attribute — and for `<sparkle:minimumSystemVersion sparkle:os="win">` to work, both Sparkle and WinSparkle would need to be updated.
